### PR TITLE
Ensure our socket one-shot 'free' event handler runs first

### PR DESCRIPTION
**Pull request by evantorrie**
This fixes a problem in any version of Node after v0.12 where the
built-in http(s) Agent code creates the socket and attaches a 'free'
handler before our ClientRequest.onSocket monkeypatch installs its own
'free' event handler.

When the Agent's free handler runs, it `null`s the socket's _httpMessage
which our 'free' handler relies upon to clear the associated timer.

Conditions to reproduce:
* out_req_timeout is set to a number less than the other
  timeouts (incoming and global),
* the process is using Agent keepalive to handle the socket pool

Symptoms:
* initial request to a destination occurs at time t
* subsequent requests to the same destination occur on the same socket
* at time `t + out_req_timeout`, the socket is destroyed when our code
  calls message.abort() on the initial request
* Any in-progress request at `t + out_req_timeout` on that socket is
  dropped on the floor.

This fix uses node's EventEmitter prependOnceListener() (available
since Node 6) to ensure our handler runs before the Agent's free
handler, thereby leaving the socket's _httpMessage pointing to the
correct originating request when our `free` handler runs.

For Node v4.x, we disable all timeouts if the request appears to be marked as keepAlive.  In addition, we have dropped support in this version for any Node earlier than v4.x

N.B.  Relying on an undocumented API ('free' handler) is likely to
bite you in the rear.  Case in point.

### CODE REVIEW

**Comment by sylvio** [comment]
+1

**Review by drewfish — APPROVED** [review]

**Comment by redonkulus** [comment]
@evantorrie does this resolve https://github.com/yahoo/node-limits/issues/10?

**Comment by evantorrie** [comment]
Yes.  I believe this will fix #10.  I guess this has been broken since they added the working `http.Agent` in node-0.12.

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-    - "0.10"
-    - "0.12"
     - "4"
```
Comment (drewfish) [inline .travis.yml]: node.js 4.x is EOL so I think we can drop it if we want

Comment (evantorrie) [reply .travis.yml]: I've managed to leave it in there, but I disable outgoing request timeouts for keepalive on Node v4.x and earlier.

**Comment by evantorrie** [comment]
@drewfish Can you rereview?

I added a functional test which tests the exact problem that Membership were having (every "nth" keepalive request being killed due to a out_req_timeout setting which was >> any individual outgoing request/response cycle, but < n * keepalive req/res time.)

You can see the functional test being added in the series of commits above (https://github.com/yahoo/node-limits/pull/15/commits/fa823c5a2e88d737be05fcc4ad04efeff1c14567) and failing.  Then the next commit fixes the problem and returns the functional test to *green*.

File: test/functional.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+            req.resume()
+                .on('end', function() {
+                    setTimeout(function() {
+                        res.writeHead(200, 'OK');
+                        res.end();
+                    }, latency);
+                });
+        };
+        self.server = http.createServer(app).listen(kPort);;
```
Comment (drewfish) [inline test/functional.js]: (double semicolon)

File: test/functional.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        // Make multiple consecutive requests, each one taking kLatency milliseconds
+        // such that kLatency << out_req_timeout, but kNumRequests * kLatency > out_req_timeout
+        async.timesSeries(kNumRequests, function(i, next) {
+            var rq = { emit: sinon.fake() };
+            var rs = { write: sinon.fake(), writeHead: sinon.fake(), end: sinon.fake() };
+
+            // Emulate node-limits middleware running on each incoming request
+            // where it monkey patches and sets various timeouts for outgoing ClientRquests
+            testee(rq, rs, function(err) {
+                if (err) { console.log('Error in node-limits client handler'); }
```
Comment (drewfish) [inline test/functional.js]: should this err cause the test to fail (somehow) ?

File: test/functional.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                        next(e, req);
+                    }).on('abort', function(e) {
+                        next(e, req);
+                    });
+
+                    req.end();
+                })(rq, rs);
+
+            });
+        }, function(err, reqs) {
```
Comment (drewfish) [inline test/functional.js]: should an err cause the tests to fail (somehow) ?

**Comment by evantorrie** [comment]
I added a call to `done(err)` which will result in mocha failing the test if any error is propagated from the async timesSeries

**Review by drewfish — APPROVED** [review]

**Pull request merged**